File: _site/python_solutions/1071.greatest-common-divisor-of-strings.js
/*
 * @lc app=leetcode id=1071 lang=javascript
 *
 * [1071] Greatest Common Divisor of Strings
 *
 * https://leetcode.com/problems/greatest-common-divisor-of-strings/description/
 *
 * algorithms
 * Easy (48.82%)
 * Total Accepted:    3.1K
 * Total Submissions: 6.4K
 * Testcase Example:  '"ABCABC"\n"ABC"'
 *
 * For strings S and T, we say "T divides S" if and only if S = T + ... + T  (T
 * concatenated with itself 1 or more times)
 * 
 * Return the largest string X such that X divides str1 and X divides str2.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: str1 = "ABCABC", str2 = "ABC"
 * Output: "ABC"
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: str1 = "ABABAB", str2 = "ABAB"
 * Output: "AB"
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: str1 = "LEET", str2 = "CODE"
 * Output: ""
 * 
 * 
 * 
 * 
 * Note:
 * 
 * 
 * 1 <= str1.length <= 1000
 * 1 <= str2.length <= 1000
 * str1[i] and str2[i] are English uppercase letters.
 * 
 */
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */


var bisect_left = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = ~~((i + j) / 2);
        if (nums[mid] < target)
            i = mid + 1;
        else
            j = mid;
    }
    return i;
}

var bisect_right = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = Math.ceil((i + j) / 2);
        if (nums[mid] > target)
            j = mid - 1;
        else
            i = mid;
    }
    return j;
}

var two_d_array = function(m, n) {
    return new Array(m).fill(0).map(() => new Array(n).fill(0));
}

var reverse = function(s) {
    return [...str].reverse().join('');
}


// simple print
var print = function(a) {
    console.log(a);
}

var say = function(a) {
    console.log(a)
};

// For counting elements in a list 
var counter = function(array_or_string) {
    var dict = {};
    for (let i = 0; i < array_or_string.length; i++) {
        let e = array_or_string[i];
        dict[e] = dict[e] ? dict[e] += 1 : 1;
    }
    return dict;
}

var sum = function(array) {
    return array.reduce(function(a, b) {
        return a + b;
    }, 0);
}

var isodd = function(n) {
    return n % 2 == 1;
}

var iseven = function(n) {
    return n % 2 == 0;
}

var list = function(size, value) {
    return Array(size).fill(value);
}


var len = function(str_or_array) {
    return str_or_array.length;
}

var sorted = function(arr) {
    return arr.sort((a, b) => (a - b));
}

// print out a Map 
function printMap(m) {
    m.forEach(function(v, k) {
        console.log(k + " " + v);
    })
}

var gcd = function(a, b) {
    return b == 0 ? a : gcd(b, a % b);
}

var divide = function(s, z) {
    for (let i = 0; i < len(s); i++)
        if (s[i] != s[i % z]) return false;
    return true;
}

var gcdOfStrings = function(s1, s2) {
    var m = len(s1),
        n = len(s2),
        z = gcd(m, n);
    if (divide(s1, z) && divide(s2, z) && s1.substring(0, z) == s2.substring(0, z))
        return s1.substring(0, z);
    return ""
};

print(gcdOfStrings("ABAB", "ABABAB"))

File: _site/python_solutions/1071.greatest-common-divisor-of-strings.rb
#
# @lc app=leetcode id=1071 lang=ruby
#
# [1071] Greatest Common Divisor of Strings
#
# https://leetcode.com/problems/greatest-common-divisor-of-strings/description/
#
# algorithms
# Easy (48.82%)
# Total Accepted:    3.1K
# Total Submissions: 6.4K
# Testcase Example:  '"ABCABC"\n"ABC"'
#
# For strings S and T, we say "T divides S" if and only if S = T + ... + T  (T
# concatenated with itself 1 or more times)
#
# Return the largest string X such that X divides str1 and X divides str2.
#
#
#
# Example 1:
#
#
# Input: str1 = "ABCABC", str2 = "ABC"
# Output: "ABC"
#
#
# Example 2:
#
#
# Input: str1 = "ABABAB", str2 = "ABAB"
# Output: "AB"
#
#
# Example 3:
#
#
# Input: str1 = "LEET", str2 = "CODE"
# Output: ""
#
#
#
#
# Note:
#
#
# 1 <= str1.length <= 1000
# 1 <= str2.length <= 1000
# str1[i] and str2[i] are English uppercase letters.
#
#
# @param {String} str1
# @param {String} str2
# @return {String}
def gcd(a, b)
  b == 0 ? a : gcd(b, a.modulo(b))
end

def divide(s, m)
  0.upto(s.size - 1).each do |i|
    return false if s[i] != s[i % m]
  end
  true
end

def gcd_of_strings(s1, s2)
  m = gcd(s1.size, s2.size)
  return s1[0..m - 1] if divide(s1, m) && divide(s2, m) && s1[0..m - 1] == s2[0..m - 1]

  ''
end

str1 = 'ABABAB'
str2 = 'AB'
p gcd_of_strings(str1, str2)
